### ISSUE TITLE
ca65: cheap local label scope vs. variable definition

### ISSUE THREAD

**Issue by fadden** [cc65/cc65#938]
A question about "cheap local labels"...

This is not expected to work:

    LAB1:   nop
            beq     @LAB2
    FOO     nop
    @LAB2:  nop

because an instruction with a non-local label sits between the definition of and reference to `@LAB2`.

However, I was surprised to discover that defining a constant also ends the scope:

    LAB1:   nop
            beq     @LAB2
    FOO     =       $1234
    @LAB2:  nop

as does defining a numeric variable:

    LAB1:   nop
            beq     @LAB2
    FOO     .set    $1234
    @LAB2:  nop

The latter case is of interest to me because SourceGen now allows the definition of local variable tables, which allow the user to redefine zero-page address labels and 65816 stack-relative constants to match what a subroutine is doing.  The local variable tables are interfering with the automatic label localization.

I was wondering if this was intentional.

**Comment by greg-king5** [comment]
> I was wondering if this was intentional.

It's intentional.  This:
```asm
LAB1:   nop
        beq     @LAB2
FOO := *
        nop
@LAB2:  nop
```
is the same as:
> ```asm
> LAB1:   nop
>         beq     @LAB2
> FOO:    nop
> @LAB2:  nop
> ```

**Comment by mrdudz** [comment]
the programmcounter assignment should be handled as a special case, imho - its the one and only case where breaking the scope is necessary

**Comment by greg-king5** [comment]
> the programmcounter assignment should be handled as a special case, imho -- it's the one and only case where breaking the scope is necessary.

OK, I will mark it as a bug.  I have no idea how easy or hard it will be to fix.

**Comment by mlauke** [comment]
can confirm the cheap local label scope is "closed" or ends with a variable definition inside a macro or without the macro at all.

```
@lp:
    sys_delay_ms 1
    dec 
    bne @lp
```
with macro sys_delay_ms defined as
```
.macro sys_delay_us micros
    .local @0,@1,@2
    .local cnt
    cnt = (micros * clockspeed / 6)
    ....
```
just the local variable assignment will end the label scope. we end up with assembly error regarding the cheap local label **@lp**
``` 
delay.s(21): Error: Symbol '@lp' is undefined
make: *** [Makefile:92: ../libwrk/steckschwein-clib/delay.o] Error 1
````

**Comment by bbbradsmith** [comment]
This issue is strongly related to: #505

**Comment by adiee5** [comment]
> can confirm the cheap local label scope is "closed" or ends with a variable definition inside a macro or without the macro at all.
> 
> ```
> @lp:
>     sys_delay_ms 1
>     dec 
>     bne @lp
> ```
> 
> with macro sys_delay_ms defined as
> 
> ```
> .macro sys_delay_us micros
>     .local @0,@1,@2
>     .local cnt
>     cnt = (micros * clockspeed / 6)
>     ....
> ```

Yep, I also had an issue regarding labels inside macros closing a Cheap scope

**Comment by mrdudz** [comment]
I changed this to feature request - because it is not a bug. It works as advertised (and documented)

That a different behaviour might be wanted and/or desirable is another question :)
